amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/logging/DiagnosticEvent.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates.
+ * Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package software.amazon.kinesis.coordinator.logging;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.NotImplementedException;
+
+import java.util.function.Consumer;
+
+/**
+ * This class is used to log state information periodically via a background thread in the KCL.
+ */
+@Slf4j
+@ToString
+public abstract class DiagnosticEvent<T extends DiagnosticEvent> {
```

Comment:
There are too many levels of inheritance. Avoid using abstract classes if possible. This could be done as an interface and a implementation.

Reply:
@ashwing do you have any thoughts on this note? The design we originally discussed used inheritance but after talking to Sahil it sounds like the KCL 2.X strongly prefers interfaces

Reply:
We can use interfaces and have a visitor pattern to enforce customer to handle all the events we generate.

Reply:
Refactored to use visitor pattern for logging behavior handling, rather than inheritance from abstract class